### ISSUE TITLE
Support functools.partial

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#1536]
The type stubs for `partial` can't describe its behavior in a useful way, so type checkers special case it. We should too, so that the return type of `partial` tracks the full underlying signature and the partial arguments provided, and we can later combine the partial arguments with newly-supplied arguments and check them against the wrapped signature. For example:

```py
from functools import partial

def f(a: int, b: str) -> bool: ...

g = partial(f, 1)

g(1, "foo") # error: too many arguments
g(1) # error: int not assignable to str
g("foo") # good
```

**Comment by davidbrochart** [comment]
Could it support classes as well? For instance, this code fails:
```py
from functools import partial

class Foo:
    def __init__(self, a):
        pass

def bar(factory: type[Foo]):
    pass

Foo1 = partial(Foo, 1)
bar(Foo1)
# error: Argument 1 to "bar" has incompatible type "partial[Foo]"; expected "type[Foo]"  [arg-type]
```
While `bar(Foo)` works fine.

**Comment by NeilGirdhar** [comment]
> While `bar(Foo)` works fine.

I think that Ty is doing the right thing here by telling you that `partial[Foo]` is _not_ a type. What you want to pass in is a factory.  A factory is something you can call to return an object.  Therefore, you should annotate it like that:
```
def bar(factory: Callable[[], [Foo]]) -> None:
```
This was discussed in [detail here](https://github.com/python/mypy/issues/4717#issuecomment-1376284599).

Hope you agree! 😄

**Comment by davidbrochart** [comment]
Right, thanks @NeilGirdhar !
